pamelalozano/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -14,31 +14,99 @@
 
 package com.google.sps.servlets;
 
+import java.util.Enumeration;
+import com.google.appengine.api.datastore.Cursor;
+import com.google.appengine.api.datastore.QueryResultList;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
 import com.google.appengine.api.datastore.DatastoreService;
 import com.google.appengine.api.datastore.DatastoreServiceFactory;
 import com.google.appengine.api.datastore.Entity;
 import java.util.ArrayList; 
+import java.util.List; 
 import java.util.Date; 
 import java.io.IOException;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
+//ToDo: Separate doPost and doGet in different files
+
 /** Servlet that returns some example content.*/
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
-    
-  ArrayList<String> comments = new ArrayList<>();  
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    FetchOptions fetchOptions = FetchOptions.Builder.withLimit(3);
+
+     String startCursor = request.getParameter("cursor");
+     String next = request.getParameter("next"); //if back this is null
+     String back = request.getParameter("back"); //if next this is null
+
+    //If the request is for a change of page
+    if (startCursor != null && next != null) {
+        
+        //Next 3 comments
+      fetchOptions.startCursor(Cursor.fromWebSafeString(startCursor));
+    
+    } else if (startCursor != null && back != null) {
+        
+        //Back to the first 3 comments 
+        //I couldn't figure out how to get the getStartCursor()
```

Comment:
You may be able to use QueryResultIterator instead QueryResultList to get both the start and end cursors? https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/QueryResultIterator.html

Reply:
Yeah I spent a long time trying to implement this but i don't know why I get the error "Class query doesn't take parameters" I guess because the query is declared like this (according to https://cloud.google.com/datastore/docs/concepts/queries):
Query<Entity> query = Query.newEntityQueryBuilder()

But it didn't let me so I tried to look for another import I may need but I didn't find anything or what I found didn't work. 
The class Query and StructuredQuery seem contain the getStartCursor method:
https://developers.google.com/resources/api-libraries/documentation/datastore/v1/java/latest/com/google/api/services/datastore/v1/model/Query.html
https://googleapis.dev/java/google-cloud-clients/0.90.0-alpha/com/google/cloud/datastore/StructuredQuery.html

Let me know if you find out why this is happening :)

Reply:
Not sure but maybe you want to try setting the start cursor the same, but call `Query.reverse()`?https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/Query.html#reverse--. Results might be sorted in the opposite way though.

I think you're looking at the docs for the newer API, which they're not having you use for some reason. You want to look through https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/package-summary

Reply:
Ohh okay perfect I think I'll try that when I'm done with week 4 so I don't fall behind. Thank you!